### PR TITLE
Fix ISA string parsing

### LINKED ISSUE #132: Use `Array{UInt8}(undef, len)` instead of `repeat(" ", len)`
Julia strings are defined to be immutable. So this is technically not valid

https://github.com/JuliaGPU/AMDGPU.jl/blob/43124078c1c5b6359967877d1f1dab0ba4acdf2b/src/agent.jl#L196

`String(buf)` makes a zero copy `String` from an `Array`

### PR DESCRIPTION
**Pull request by jpsamaroo**
Fixes #132

### CODE REVIEW

**Comment by vchuravy** [comment]
Good to merge? Can confirm that this is necessary on the MI50 system at ANL and there yields: 

```
julia> AMDGPU.default_isa(AMDGPU.default_device())                                                                                                                                                                                            
"gfx906"                                                                                                                                                                                                                                      
```

**Comment by jpsamaroo** [comment]
Sorry, I meant to ask for someone with such a system to test that codegen+launch also works. If you can confirm it does work, I'll merge this.

**Comment by vchuravy** [comment]
As noted on slack currently fails with:

```
julia> fill!(x, 0.0)                                                                                                                                                                                                                          
ERROR: HSA error (code #4109, HSA_STATUS_ERROR_INCOMPATIBLE_ARGUMENTS: The arguments passed to a functions are not compatible.)                                                                                                               
Stacktrace:                                                                                                                                                                                                                                   
  [1] check                                                                                                                                                                                                                                   
    @ ~/.julia/dev/AMDGPU/src/error.jl:34 [inlined]                                                                                                                                                                                           
  [2] |>                                                                                                                                                                                                                                      
    @ ./operators.jl:858 [inlined]                                                                                                                                                                                                            
  [3] HSAExecutable(agent::HSAAgent, data::Vector{UInt8}, symbol::String; globals::Vector{Pair{Symbol, _A} where _A})
    @ AMDGPU ~/.julia/dev/AMDGPU/src/executable.jl:55
  [4] create_executable(::AMDGPU.DeviceRuntime, device::AMDGPU.RuntimeDevice{HSAAgent}, entry::String, obj::Base.CodeUnits{UInt8, String}; globals::Vector{Pair{Symbol, _A} where _A})
    @ AMDGPU ~/.julia/dev/AMDGPU/src/runtime.jl:98
  [5] #create_executable#28
    @ ~/.julia/dev/AMDGPU/src/runtime.jl:83 [inlined]
  [6] rocfunction_link(job::GPUCompiler.CompilerJob, compiled::NamedTuple{(:obj, :entry, :globals), Tuple{String, String, Vector{Pair{Symbol, _A} where _A}}})
    @ AMDGPU ~/.julia/dev/AMDGPU/src/execution.jl:331
```

File: src/agent.jl [hunk 1]
```diff
@@ -189,15 +189,18 @@ function isas(agent::HSAAgent)
     HSA.agent_iterate_isas(agent.agent, func, isas) |> check
     isas[]
 end
-function get_first_isa_string(agent::HSAAgent)
-    isa = first(isas(agent))
+const isa_regex = r"([a-z]*)-([a-z]*)-([a-z]*)--([a-z0-9]*)([a-zA-Z0-9+]*)"
```
Comment (vchuravy) [inline src/agent.jl:192]: This might not be thread-safe

Comment (jpsamaroo) [reply src/agent.jl:192]: So, generate it on-the-fly?

**Pull request closed without merging**